### ISSUE TITLE
Incremental parsing bugs when using non-terminal extras

### ISSUE THREAD

**Issue by Pyromuffin** [tree-sitter/tree-sitter#633]
Is it expected that two strings would parse the differently, after being edited to be identical strings?
I'm finding that the following strings parse differently after edited:

1)
```
main :: () {
    print();
}
// comment
#import "Basic";
```
(source_file (named_block_decl (identifier) (function_definition (block (func_call (identifier) (argument_list)) (block_end)))) (inline_comment) (import_statement name: (string_literal)))

2)
```
main :: () {
    print();
}
// comment
/#import "Basic";
```
(source_file (named_block_decl (identifier) (function_definition (block (func_call (identifier) (argument_list)) (block_end)))) (inline_comment) (ERROR) (import_statement name: (string_literal)))

3)
```
main :: () {
    print();
}
// comment
/ #import "Basic";
```
(source_file (named_decl (identifier) (binary_expression (function_definition (block (func_call (identifier) (argument_list)) (block_end))) (inline_comment) (ERROR) (string_literal))))

4)
```
main :: () {
    print();
}
// comment
/#import "Basic";
```
(source_file (named_decl (identifier) (binary_expression (function_definition (block (func_call (identifier) (argument_list)) (block_end))) (inline_comment) (ERROR) (string_literal))))

Basically, in 2) after adding a "/" before the "#import" line, the first time it detects an error and still finds an import statement. 

Then in 3) I add a space after the "/"  After adding a space, it tries to parse it as a binary expression where we are dividing  the function definition by the string literal (which is not allowed, but an understandable way to parse this error). 

In 4) I remove the space after the "/" so now the string is identical to 2), but the string is parsed differently. It is still parsing as if it were a binary expression. At this point, I think the parser is in a weird state, as further editing back to 1) still results in an error, which I have not shown here.

I'm open to the possibility that I am performing edits incorrectly, but I am fairly straightforwardly using the the edit helper in https://github.com/tree-sitter/tree-sitter/blob/a04d688d3895e17bf6111722aa57b7148d8f9df1/cli/src/parse.rs#L206 (adapted to c++).

It was a lot of work to cut this down to the minimal repro case, so I would greatly appreciate further direction on how to debug this or if this is the expected behavior.

Here is the minimal grammar for the repro, I hope it is not too difficult to understand.

```module.exports = grammar({
    name: 'test',
    extras: $ => [$.inline_comment, /[\s]+/],

    rules: {

      source_file: $ => repeat(
          $._statement,
        ),

      _expression: $ => choice(
        $.string_literal,
        $.binary_expression,
        $.identifier,
        $.func_call,
        $.function_definition,
      ),

      _statement: $ => choice(
        $.block,
        $._expression_statement,
        $.named_block_decl,
        $.named_decl,
        $.import_statement,
      ),
  
      _expression_statement: $ => seq( $._expression, ";"),
  
      inline_comment: $ => seq('//', /.*/),

      string_literal: $ => seq(
        '"',
        repeat(choice(/./, /\\./, "//", "/*", "*/")),
        '"'
      ),

      import_statement: $ =>  seq(
        '#import',
        field("name", $.string_literal),
        ";"
      ),


        argument_list: $ => prec(1, seq( // precedence over function definition
          "(",
          ")",
        )),

        func_call: $ =>seq(
          $._expression,
          $.argument_list,
        ),
          
        function_definition : $ => seq(
          "()",
          $.block,
        ),
    

      block: $ => seq(
        '{',
        repeat($._statement),
        $.block_end,
      ),
      block_end: $ => "}",

      named_block_decl: $ => prec(1, seq( // precedence over named decl
        $._expression,
        "::",
        $.function_definition
      )),
      
      named_decl: $ => seq(
        $._expression,
        "::",
        $._expression,
        ";",
      ),

      binary_expression: $ => choice(
        prec.left(2, seq($._expression, '/', $._expression)),
      ),

      identifier: $ => /[a-zA-Z_][a-zA-Z_0-9]*/,
    }
  }
);
```

Thanks!
Kelly

**Comment by Pyromuffin** [comment]
Another interesting tangle to this is that calling ts_parser_parse twice in a row, using the same "old_tree" reverts the parsing of 4) back to the parsing of 2).

**Comment by maxbrunsfeld** [comment]
This is a bug.

The intended behavior is that `ts_parser_parse` should produce the same tree after an edit as it would when parsing from scratch, even in the case of errors. In the test suite, we have [fewer tests](https://github.com/tree-sitter/tree-sitter/tree/master/test/fixtures/error_corpus) that enforce this property in the case of *errors* than in the case of valid code, so I'm not totally surprised we're seeing this.

> At this point, I think the parser is in a weird state, as further editing back to 1) still results in an error

I think that would be a more severe bug, but I haven't been able to reproduce that. As far as I could tell, changing the code back to 1) resulted in the correct tree.

FWIW, I am reproducing these at the command line using the `tree-sitter parse --edit` flag. It allows you to parse a file, then perform a text change and re-parse. You can perform a series of edits by supplying the flag multiple times, like this:

```sh
# Insert '/' in example 1, producing example 2
tree-sitter parse example-1 --edit '4,0 0 /'
```

```sh
# Remove a character from example 3, producing example 4
tree-sitter parse example-3 --edit '4,1 1'
```

I will try and debug this, but I don't have an ETA, because it usually takes some time and focus to investigate these incremental parsing bugs.

Thanks for spending the time to write up the repro report.

**Comment by Pyromuffin** [comment]
Thanks for the quick response maxbrunsfeld, the case that causes incorrect parsing on returning to 1) is as follows:
start with:
```
main :: () {
    print();
}
// comment
///#import "Basic";
```

or alternatively add the three '/' in sequential edits to the import line, it doesn't make a difference.
and then delete the three '/'. The resulting parse is:

(source_file (named_decl (identifier) (function_definition (block (func_call (identifier) (argument_list)) (block_end))) (inline_comment) (MISSING ";")) (import_statement name: (string_literal)))

as you might imagine, this was the initial problem I stumbled across, and reduced it to the more minimal case mentioned in my first post, however, if it were just incorrect error recovery, that would be somewhat more manageable. The fact that it causes an incorrect parse on returning to valid code is much more problematic.

Thanks,
Kelly

**Comment by maxbrunsfeld** [comment]
Ok, this exposed some incremental logic that was blatantly missing, fixed on this branch: #636.

I'm going to wait to merge that PR until I've got proper test coverage in place, but in the meantime you may want to test that your issues are solved.

I really appreciate the detailed report, it is really helpful for the project!

**Comment by Pyromuffin** [comment]
That branch does indeed fix the problem indicated in my last comment, but I found another case of inconsistent parsing with editing, which requires just a very slightly more complicated grammar. Note the addition of $.compiler_directive_simple:

```module.exports = grammar({
  name: 'test',
  extras: $ => [$.inline_comment, /[\s]+/, $.compiler_directive_simple],

  rules: {


    source_file: $ => repeat(
        $._statement,
      ),

    _expression: $ => choice(
      $.string_literal,
      $.binary_expression,
      $.identifier,
      $.func_call,
      $.function_definition,
    ),

    _statement: $ => choice(
      $.block,
      $._expression_statement,
      $.named_block_decl,
      $.named_decl,
      $.import_statement,
    ),

    _expression_statement: $ => seq( $._expression, ";"),

    inline_comment: $ => seq('//', /.*/),

    string_literal: $ => seq(
      '"',
      repeat(choice(/./, /\\./, "//", "/*", "*/")),
      '"'
    ),

    import_statement: $ =>  seq(
      '#import',
      field("name", $.string_literal),
      ";"
    ),

    compiler_directive_simple: $ => /\#[a-zA-Z_][a-zA-Z_0-9]*/,


      argument_list: $ => prec(1, seq( // precedence over function definition
        "(",
        ")",
      )),

      func_call: $ =>seq(
        $._expression,
        $.argument_list,
      ),
        
      function_definition : $ => seq(
        "()",
        $.block,
      ),
  

    block: $ => seq(
      '{',
      repeat($._statement),
      $.block_end,
    ),
    block_end: $ => "}",

    named_block_decl: $ => prec(1, seq( // precedence over named decl
      $._expression,
      "::",
      $.function_definition
    )),
    
    named_decl: $ => seq(
      $._expression,
      "::",
      $._expression,
      ";",
    ),

    binary_expression: $ => choice(
      prec.left(2, seq($._expression, '/', $._expression)),
    ),

    identifier: $ => /[a-zA-Z_][a-zA-Z_0-9]*/,
  }
}
);
```
then parse the following:

1)
```
main :: () {
    print();
}
// comment
#import "Basic";
```
(source_file (named_block_decl (identifier) (function_definition (block (func_call (identifier) (argument_list)) (block_end)))) (inline_comment) (import_statement name: (string_literal)))

2)
```
main :: () {
    print();
}
// comment
/#import "Basic";
```
(source_file (named_decl (identifier) (binary_expression (function_definition (block (func_call (identifier) (argument_list)) (block_end))) (inline_comment) (compiler_directive_simple) (string_literal))))

3)
```
main :: () {
    print();
}
// comment
#import "Basic";
```
(source_file (named_block_decl (identifier) (function_definition (block (func_call (identifier) (argument_list)) (block_end)))) (inline_comment) (compiler_directive_simple) (string_literal))

note that in 3) the import statement is now being parsed as a compiler_directive_simple, instead of an import statement as in 1).

I've tested this same example on master, and 3) parses correctly as:
(source_file (named_block_decl (identifier) (function_definition (block (func_call (identifier) (argument_list)) (block_end)))) (inline_comment) (import_statement name: (string_literal)))

So that makes me thinking that there was something breaking introduced in this new branch.

Thanks,
Kelly

**Comment by maxbrunsfeld** [comment]
> I found another case of inconsistent parsing with editing, which requires just a very slightly more complicated grammar.

Ok, I think this is a separate bug. I think the workaround for this would be to make your `comment` rule into a single token, instead of a parent node that has two children (`//` and `.*`). To do this, you just wrap it with the `token` function like this:

```js
inline_comment: $ => token(seq('//', /.*/)),
```

This will also make your parser faster and smaller. You shouldn't *need* to do this to get correct incremental parsing, but [non-terminal extras](https://github.com/tree-sitter/tree-sitter/pull/469) is a somewhat newer feature, and we haven't used it in that many places in the grammars that we maintain.

> Interestingly, 2) parses as ...
> So that makes me think that there was something breaking introduced in this new branch.

I think that the new behavior is correct here. Example 2 gives the same syntax tree whether you're parsing from scratch, or editing example 1. Before, that was not the case.

**Comment by maxbrunsfeld** [comment]
Also, somewhat unrelated, another performance win would be to make your `string_literal` rule into a single token. Right now, it is creating a separate token corresponding to each regex match - essentially a node per character.

```js
    string_literal: $ => token(seq(
      '"',
      repeat(choice(/./, /\\./, "//", "/*", "*/")),
      '"'
    )),
```

Sometimes, you might want a separate node for the delimiters, or things like escape sequences. You can check out the way we handle this [in the C grammar](https://github.com/tree-sitter/tree-sitter-c/blob/99151b1e9293c9e025498fee7e6691e1a52e1d03/grammar.js#L917-L924) for an example of how to deal with this.

Basically, we create a separate node for each *continuous run* of normal characters, instead of one for each individual character. `token(prec(1, STUFF))` is the pattern for preventing string contents from being parsed as comments or other `extras` tokens.

**Comment by Pyromuffin** [comment]
Thanks for the tip, I'll look into the C grammar. My parser is hideously large at the moment (nearing 10 mb) but the language I am parsing has a lot of edge cases, and attempts to simplify have been somewhat countered by that. I haven't been using the token() rule, so perhaps that can help!

Thanks,
Kelly

**Comment by Pyromuffin** [comment]
Adding that token rule to inline_comment appears to have fixed it, even in the full grammar. I'm unable to break the parser as far as I can tell for now.

**Comment by maxbrunsfeld** [comment]
Ok, I found that the first bug you reported was easy to cover with tests; I just had to add our Ruby parser to our randomized test suite. The second one will be a bit more work to get test coverage for, and I'm not sure of the fix yet, so I've opened a separate issue for it ☝️ .

I believe both bugs were ultimately caused by using the non-terminal version of `line_comment`, so I don't think either will affect you now that you're using `token` in `line_comment`. It might be worth upgrading to the latest commit on master anyway.
